File: app/src/main/java/miles/diary/ui/Predrawer.java
package miles.diary.ui;

import android.view.View;
import android.view.ViewTreeObserver;

/**
 * Created by mbpeele on 2/2/16.
 */
public abstract class PreDrawer<T extends View> {

    public PreDrawer(final T view) {
        final ViewTreeObserver viewTreeObserver = view.getViewTreeObserver();
        viewTreeObserver.addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                if (viewTreeObserver.isAlive()) {
                    viewTreeObserver.removeOnPreDrawListener(this);
                } else {
                    view.getViewTreeObserver().removeOnPreDrawListener(this);
                }

                notifyPreDraw(view);
                return true;
            }
        });
    }

    public abstract void notifyPreDraw(T view);
}

File: app/src/main/java/miles/diary/ui/PreDrawer.kt
package miles.diary.ui

import android.view.View
import android.view.ViewTreeObserver

import java.lang.ref.SoftReference

/**
 * Created by mbpeele on 2/2/16.
 */
fun <T : View> T.addPreDrawer(receiver: (T) -> (Unit)) {
    val softReference = SoftReference(this)

    val viewTreeObserver = viewTreeObserver
    viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
        override fun onPreDraw(): Boolean {
            if (viewTreeObserver.isAlive) {
                viewTreeObserver.removeOnPreDrawListener(this)
            } else {
                this@addPreDrawer.viewTreeObserver.removeOnPreDrawListener(this)
            }

            val reference = softReference.get()
            return if (reference == null) {
                false
            } else {
                receiver.invoke(reference)
                true
            }
        }
    })
}
